Update JCasC Drift detector project (#4864)
jenkins-infra/jenkins.io: content/projects/gsoc/2022/project-ideas/JCasC-drift-detector.adoc

mentors:
- "jmMeessen"
links:
   emailThread: https://community.jenkins.io/t/gsoc-2022-project-proposal-jenkins-configuration-as-code-jcasc-drift-detector/1325
---

=== Context 

> A well designed Configuration Process will only allow updating Jenkins via Configuration as Code. 

Reviewer: ```suggestion
A well designed Configuration Process only allows updating Jenkins via Configuration as Code. 
```

Author: incorporated. Thank you for suggesting.

> It will thus periodically and "blindly" overwrite the controller to make sure that the configuration code is equivalent to the running one. 

Reviewer: ```suggestion
Consequently, it periodically and "blindly" overwrites the controller to make sure that the configuration code is equivalent to the running one. 
```

Author: incorporated. Thank you for suggesting.

This is the recommended practice (best if the interval is kept very short).

But experience shows that compromises need to be made. 

> These exceptions range from interactively validating touchy configuration change on Production, to allowing a developer to experiment the best parameters for a feature, or to an emergency interactive configuration change.

Reviewer: ```suggestion
These exceptions range from interactively validating touchy configuration changes on Production, to allowing a developer to experiment to determine the best parameters for a feature, or to an emergency interactive configuration change.
```

Author: incorporated. Thank you for suggesting.

This is typically the case when trying to reach the ideal Configuration Process maturity level. 

> Another use case is to be able to compare the configuration of two Jenkins controllers (ex: Test vs. Production) 

Reviewer: ```suggestion
Another use case is to compare the configuration of two Jenkins controllers (ex: Test vs. Production) 
```

Author: incorporated. Thank you for suggesting.

=== Project proposal

This project intents to create the tooling, similar to the terraform plan command, to:

> * Show what configuration elements would be created or changed if the supplied JCasC file was applied to the given Jenkins system.

Reviewer: ```suggestion
* Show what configuration elements would be created or changed if the supplied JCasC file was applied to the specified Jenkins system.
```

Author: incorporated. Thank you for suggesting.

** Ability to configure the details displayed for sensitive elements (credentials, sensitive configuration, etc)

> ** Ability to opt out above mentioned sensitive elements. These elements may be absent or encrypted in the supplied JCasC file.

Reviewer: ```suggestion
** Ability to opt out of updating or displaying these sensitive elements. These elements may be absent or encrypted in the supplied JCasC file.
```
I don't know if this is correct but I think we need something -- just opting out of these elements doesn't make sense.

Author: Your suggestion clarified what I had in mind. But more thought on this is required in the early stages of that proposed project.

* Generate a JCasC yaml with only the detected delta.

> * Give a boolean result whether the controller's configuration is equivalent to the supplied JCasC configuration (did the configuration drift?)

Reviewer: ```suggestion
* Give a boolean result indicating whether the controller's configuration is equivalent to the supplied JCasC configuration (did the configuration drift?)
```

Author: incorporated. Thank you for suggesting.
